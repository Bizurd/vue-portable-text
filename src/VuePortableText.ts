import Vue, { CreateElement, PropType } from 'vue'
import defu from 'defu'
import { Serializers } from './types/serializers'
import { VNode } from 'vue/types/umd'
import { defaultSerializers } from './serializers'

type Blocks = any[] | any
type Block = any

const createList = (block: Block, blocks: Blocks, serializers: Serializers) => {
  const listItems = getListItems(block, blocks, serializers)

  return {
    _type: 'list',
    listItems,
    list: block.listItem,
  }
}

const getListItems = (
  block: Block,
  blocks: Blocks,
  serializers: Serializers
) => {
  const listItems = []

  const level = block.level
  const listStyleType = block.listItem
  while (block) {
    if (block.listItem === listStyleType && block.level === level) {
      listItems.push({
        _type: 'listItem',
        level: block.level,
        children: [block],
      })
    } else if (block.level > level && listItems.length) {
      const list = createList(block, blocks, serializers)

      const lastListItem = listItems[listItems.length - 1]
      lastListItem.children = [...lastListItem.children, list]
    } else {
      blocks.unshift(block)
      break
    }
    block = blocks.shift()
  }

  return listItems
}

const groupListItems = (blocks: Blocks, serializers: Serializers) => {
  // Copy the blocks array so we can modify it
  blocks = blocks.slice()

  const newBlocks = []

  while (blocks.length) {
    const block = blocks.shift()
    if (block && block.listItem) {
      newBlocks.push(createList(block, blocks, serializers))
    } else {
      newBlocks.push(block)
    }
  }

  return newBlocks
}

const renderSpan = (
  h: CreateElement,
  block: Block,
  serializers: Serializers,
  markDefs: { _key: string; _type: string; [x: string]: any }[]
) => {
  if (!block.marks || block.marks.length === 0) return block.text

  const marks = block.marks.slice().reverse()
  return marks.reduce((previousVNode: VNode | string, mark: string) => {
    let el = serializers.marks[mark]
    const markDef = markDefs.find((markDef) => markDef._key === mark)

    if (markDef) {
      el ||= serializers.marks[markDef._type]
    }

    return el
      ? h(el, { props: { block, markDef } }, [previousVNode])
      : previousVNode
  }, block.text)
}

const renderList = (
  h: CreateElement,
  block: Block,
  serializers: Serializers
) => {
  let children = []
  if (block.listItems) {
    children = block.listItems.map((listItem: Block) =>
      renderListItem(h, listItem, serializers)
    )
  }

  return h(
    block.list === 'number' ? serializers.list.number : serializers.list.bullet,
    { props: { block } },
    children
  )
}

const renderListItem = (
  h: CreateElement,
  block: Block,
  serializers: Serializers
) => {
  return h(
    serializers.listItem,
    { props: { block } },
    renderChildren(h, block, serializers)
  )
}

const renderChildren = (
  h: CreateElement,
  block: Blocks,
  serializers: Serializers
) => {
  let children = []
  if (block.children) {
    children = block.children.map((child: Block) => {
      if (child._type === 'span') {
        return renderSpan(h, child, serializers, block.markDefs || [])
      } else if (child._type === 'list') {
        return renderList(h, child, serializers)
      } else if (child._type === 'block') {
        return renderBlock(h, child, serializers)
      } else {
        return renderCustomType(h, child, serializers)
      }
    })
  }

  return children
}

const renderBlock = (
  h: CreateElement,
  block: Block,
  serializers: Serializers
) => {
  return h(
    serializers.styles[block.style],
    { props: { block } },
    renderChildren(h, block, serializers)
  )
}

const renderCustomType = (
  h: CreateElement,
  block: Block,
  serializers: Serializers
) => {
  const serializer = serializers.types[block._type]
  if (serializer)
    return h(
      serializer,
      { props: { block } },
      renderChildren(h, block, serializers)
    )

  return h(serializers.styles.normal, {}, renderChildren(h, block, serializers))
}

export default Vue.extend({
  name: 'VuePortableText',

  props: {
    as: {
      type: String,
      default: undefined,
    },
    blocks: {
      type: Array as PropType<any[] | any>,
      default: () => [],
    },
    serializers: {
      type: Object as PropType<Partial<Serializers>>,
      default: () => ({}),
    },
  },

  render(h) {
    const serializers = defu(
      this.serializers || {},
      defaultSerializers
    ) as Serializers

    const blocks = groupListItems(this.blocks, serializers)

    const children = blocks.map((block) => {
      if (block._type === 'block') {
        return renderBlock(h, block, serializers)
      } else if (block._type === 'list') {
        return renderList(h, block, serializers)
      } else {
        return renderCustomType(h, block, serializers)
      }
    })

    return h(this.as || serializers.container, {}, children)
  },
})
