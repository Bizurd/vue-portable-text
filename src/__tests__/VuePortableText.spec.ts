/* eslint-disable vue/one-component-per-file */
import { mount } from '@vue/test-utils'
import Vue from 'vue'
import VuePortableText, { defaultSerializers as ds } from '../VuePortableText'

const container = ds.container
const p = ds.styles.normal
const h1 = ds.styles.h1
const h3 = ds.styles.h3
const blockquote = ds.styles.blockquote
const strong = ds.marks.strong
const a = ds.marks.link
const ul = ds.list.bullet
const ol = ds.list.number
const li = ds.listItem

test('no blocks or serializers', () => {
  const wrapper = mount(VuePortableText)

  expect(wrapper.element).toBeEmptyDOMElement()
  expect(wrapper.element).toContainHTML(`<${ds.container}></${ds.container}>`)
})

test('serializers = null', () => {
  const wrapper = mount(VuePortableText, {
    propsData: {
      serializers: null,
    },
  })

  expect(wrapper.element).toBeEmptyDOMElement()
  expect(wrapper.element).toContainHTML(`<${ds.container}></${ds.container}>`)
})

test('as = article', () => {
  const wrapper = mount(VuePortableText, {
    propsData: {
      as: 'article',
    },
  })

  expect(wrapper.element).toBeEmptyDOMElement()
  expect(wrapper.element).toContainHTML(`<article></article>`)
})

test('serializers.container = article', () => {
  const wrapper = mount(VuePortableText, {
    propsData: {
      serializers: {
        container: 'article',
      },
    },
  })

  expect(wrapper.element).toBeEmptyDOMElement()
  expect(wrapper.element).toContainHTML(`<article></article>`)
})

test('block.normal > plain text', () => {
  const normalText = 'This should be normal text'

  const wrapper = mount(VuePortableText, {
    propsData: {
      blocks: [
        {
          style: 'normal',
          _type: 'block',
          children: [
            {
              _type: 'span',
              text: normalText,
            },
          ],
        },
      ],
    },
  })

  expect(wrapper.element).toContainHTML(
    `<${container}><${p}>${normalText}</${p}></${container}>`
  )
})

test('block.h1 > plain text', () => {
  const h1Text = 'This should be a h1'

  const wrapper = mount(VuePortableText, {
    propsData: {
      blocks: [
        {
          style: 'h1',
          _type: 'block',
          children: [
            {
              _type: 'span',
              text: h1Text,
            },
          ],
          markDefs: [],
        },
      ],
    },
  })

  expect(wrapper.element).toContainHTML(
    `<${container}><${h1}>${h1Text}</${h1}></${container}>`
  )
})

test('block.h3 > plain text + block.blockquote > plain text', () => {
  const h3Text = 'This should be a h3'
  const blockquoteText = 'This should be a blockquote'

  const wrapper = mount(VuePortableText, {
    propsData: {
      blocks: [
        {
          style: 'h3',
          _type: 'block',
          children: [
            {
              _type: 'span',
              text: h3Text,
            },
          ],
          markDefs: [],
        },
        {
          style: 'blockquote',
          _type: 'block',
          children: [
            {
              _type: 'span',
              text: blockquoteText,
            },
          ],
          markDefs: [],
        },
      ],
    },
  })

  expect(wrapper.element).toContainHTML(
    `<${container}><${h3}>${h3Text}</${h3}><${blockquote}>${blockquoteText}</${blockquote}></${container}>`
  )
})

test('block.normal > text with marks', () => {
  const plainText = 'This should be plain text'
  const strongLinkText = 'This should be a strong link'

  const wrapper = mount(VuePortableText, {
    propsData: {
      blocks: [
        {
          style: 'normal',
          _type: 'block',
          children: [
            {
              _type: 'span',
              text: plainText,
            },
            {
              _type: 'span',
              marks: ['a-key', 'strong'],
              text: strongLinkText,
            },
          ],
          markDefs: [
            {
              _key: 'a-key',
              _type: 'link',
              href: 'https://example.com',
            },
          ],
        },
      ],
    },
  })

  expect(wrapper.element).toContainHTML(
    `<${container}><${p}>${plainText}<${a} href="https://example.com"><${strong}>${strongLinkText}</${strong}></${a}></${p}></${container}>`
  )
})

test('block.normal (listItem=bullet) * 3 > plain text', () => {
  const plainText = 'This should be plain text'

  const wrapper = mount(VuePortableText, {
    propsData: {
      blocks: [
        {
          style: 'normal',
          _type: 'block',
          children: [
            {
              _type: 'span',
              text: plainText,
            },
          ],
          level: 1,
          listItem: 'bullet',
        },
        {
          style: 'normal',
          _type: 'block',
          children: [
            {
              _type: 'span',
              text: plainText,
            },
          ],
          level: 1,
          listItem: 'bullet',
        },
        {
          style: 'normal',
          _type: 'block',
          children: [
            {
              _type: 'span',
              text: plainText,
            },
          ],
          level: 1,
          listItem: 'bullet',
        },
      ],
    },
  })

  expect(wrapper.element).toContainHTML(
    `<${container}><${ul}><${li}><${p}>${plainText}</${p}></${li}><${li}><${p}>${plainText}</${p}></${li}><${li}><${p}>${plainText}</${p}></${li}></${ul}></${container}>`
  )
})

test('block.h3 (listItem=bullet) > text with marks', () => {
  const h3Text = 'This should be a h3'
  const strongLinkText = 'This should be a strong link'

  const wrapper = mount(VuePortableText, {
    propsData: {
      blocks: [
        {
          style: 'h3',
          _type: 'block',
          children: [
            {
              _type: 'span',
              text: h3Text,
            },
            {
              _type: 'span',
              marks: ['a-key', 'strong'],
              text: strongLinkText,
            },
          ],
          markDefs: [
            {
              _key: 'a-key',
              _type: 'link',
              href: 'https://example.com',
            },
          ],
          level: 1,
          listItem: 'bullet',
        },
      ],
    },
  })

  expect(wrapper.element).toContainHTML(
    `<${container}><${ul}><${li}><${h3}>${h3Text}<${a} href="https://example.com"><${strong}>${strongLinkText}</${strong}></${a}></${h3}></${li}></${ul}></${container}>`
  )
})

test('block.normal (listItem=numeric) * 3 with nested bullet list > plain text', () => {
  const plainText = 'This should be plain text'

  const wrapper = mount(VuePortableText, {
    propsData: {
      blocks: [
        {
          style: 'normal',
          _type: 'block',
          children: [
            {
              _type: 'span',
              text: plainText,
            },
          ],
          level: 1,
          listItem: 'number',
        },
        {
          style: 'normal',
          _type: 'block',
          children: [
            {
              _type: 'span',
              text: plainText,
            },
          ],
          level: 2,
          listItem: 'bullet',
        },
        {
          style: 'normal',
          _type: 'block',
          children: [
            {
              _type: 'span',
              text: plainText,
            },
          ],
          level: 1,
          listItem: 'number',
        },
      ],
    },
  })

  expect(wrapper.element).toContainHTML(
    `<${container}><${ol}><${li}><${p}>${plainText}</${p}></${li}><${li}><${ul}><${li}><${p}>${plainText}</${p}></${li}></${ul}></${li}><${li}><${p}>${plainText}</${p}></${li}></${ol}></${container}>`
  )
})

test('customType > text', () => {
  const text = 'This should be normal text'

  const CustomTypeComponent = Vue.extend({
    props: {
      block: {
        type: Object,
        required: true,
      },
    },
    render(h) {
      return h(
        'h1',
        { attrs: { 'data-type': this.block._type } },
        this.$slots.default
      )
    },
  })

  const wrapper = mount(VuePortableText, {
    propsData: {
      serializers: {
        types: {
          customType: CustomTypeComponent,
        },
      },
      blocks: [
        {
          _type: 'customType',
          children: [
            {
              _type: 'span',
              text: text,
            },
          ],
        },
      ],
    },
  })

  expect(wrapper.element).toContainHTML(
    `<${container}><h1 data-type="customType">${text}</h1></${container}>`
  )
})

test('non-existent customType > text', () => {
  const text = 'This should be normal text'

  const wrapper = mount(VuePortableText, {
    propsData: {
      blocks: [
        {
          _type: 'customType',
          children: [
            {
              _type: 'span',
              text: text,
            },
          ],
        },
      ],
    },
  })

  expect(wrapper.element).toContainHTML(
    `<${container}><${p}>${text}</${p}></${container}>`
  )
})

test('block.normal > customType', () => {
  const text = 'This should be normal text'

  const CustomTypeComponent = Vue.extend({
    props: {
      block: {
        type: Object,
        required: true,
      },
    },
    render(h) {
      return h(
        'span',
        { attrs: { 'data-type': this.block._type } },
        this.$slots.default
      )
    },
  })

  const wrapper = mount(VuePortableText, {
    propsData: {
      serializers: {
        types: {
          customType: CustomTypeComponent,
        },
      },
      blocks: [
        {
          style: 'normal',
          _type: 'block',
          children: [
            {
              _type: 'customType',
              text: text,
            },
          ],
        },
      ],
    },
  })

  expect(wrapper.element).toContainHTML(
    `<${container}><${p}><span data-type="customType">${text}</span><${p}></${container}>`
  )
})
