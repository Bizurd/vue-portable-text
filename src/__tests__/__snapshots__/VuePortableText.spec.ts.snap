// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`as = article 1`] = `<article />`;

exports[`block.h1 > plain text 1`] = `
<div>
  <h1>
    This should be a h1
  </h1>
</div>
`;

exports[`block.h3 (listItem=bullet) > text with marks 1`] = `
<div>
  <ul>
    <li>
      <h3>
        This should be a h3
        <a
          href="https://example.com"
        >
          <strong>
            This should be a strong link
          </strong>
        </a>
      </h3>
    </li>
  </ul>
</div>
`;

exports[`block.h3 > plain text + block.blockquote > plain text 1`] = `
<div>
  <h3>
    This should be a h3
  </h3>
  <blockquote>
    This should be a blockquote
  </blockquote>
</div>
`;

exports[`block.normal (listItem=bullet) * 3 > plain text 1`] = `
<div>
  <ul>
    <li>
      <p>
        This should be plain text
      </p>
    </li>
    <li>
      <p>
        This should be plain text
      </p>
    </li>
    <li>
      <p>
        This should be plain text
      </p>
    </li>
  </ul>
</div>
`;

exports[`block.normal (listItem=numeric) * 3 with nested bullet list > plain text 1`] = `
<div>
  <ol>
    <li>
      <p>
        This should be plain text
      </p>
      <ul>
        <li>
          <p>
            This should be plain text
          </p>
        </li>
      </ul>
    </li>
    <li>
      <p>
        This should be plain text
      </p>
    </li>
  </ol>
</div>
`;

exports[`block.normal > customType 1`] = `
<div>
  <p>
    <span
      data-type="customType"
    >
      This should be normal text
    </span>
  </p>
</div>
`;

exports[`block.normal > plain text 1`] = `
<div>
  <p>
    This should be normal text
  </p>
</div>
`;

exports[`block.normal > text with marks 1`] = `
<div>
  <p>
    This should be plain text
    <a
      href="https://example.com"
    >
      <strong>
        This should be a strong link
      </strong>
    </a>
  </p>
</div>
`;

exports[`block.normal > text with nonexistent marks 1`] = `
<div>
  <p>
    This should be plain textThis should be still be plain text
  </p>
</div>
`;

exports[`customType > text 1`] = `
<div>
  <h1
    data-type="customType"
  >
    This should be normal text
  </h1>
</div>
`;

exports[`no blocks or serializers 1`] = `<div />`;

exports[`non-existent customType > text 1`] = `
<div>
  <p>
    This should be normal text
  </p>
</div>
`;

exports[`serializers = null 1`] = `<div />`;

exports[`serializers.container = article 1`] = `<article />`;
